File: app/src/main/java/cedule/app/data/dao/TaskDAO.java
package cedule.app.data.dao;

import androidx.room.Dao;
import androidx.room.Query;

import java.util.List;

import cedule.app.data.entities.Task;

@Dao
public interface TaskDAO {
    @Query("SELECT * FROM tasks")
    List<Task> getAll();

    @Query("SELECT * FROM tasks WHERE category=:category")
    List<Task> getByCategory(Integer category);

    @Query("UPDATE tasks SET title=:title, category=:cat, startDate=:startDate, startTime=:startTime, isDone=:isDone, isNotify=:isNotify, note=:note WHERE id=:taskId")
    void update(int taskId, String title, Integer cat, Long startDate, Integer startTime,
                Integer isDone, Integer isNotify, String note);

    @Query("INSERT INTO tasks (title, category, startDate, startTime, isDone, isNotify, note) " +
            "VALUES (:title, :cat, :startDate, :startTime, :isDone, :isNotify, :note)")
    void add(String title, Integer cat, Long startDate, Integer startTime,
             Integer isDone, Integer isNotify, String note);

    @Query("SELECT * FROM tasks WHERE id=:id LIMIT 1")
    Task getById(int id);

    @Query("DELETE FROM tasks WHERE id IN (:id)")
    void discard(List<Integer> id);

    @Query("UPDATE tasks SET isDone=:isDone WHERE id=:id")
    void updateStatus(Integer id, Integer isDone);

    @Query("SELECT * FROM tasks ORDER BY title DESC")
    List<Task> getInNameDesc();

    @Query("SELECT * FROM tasks ORDER BY startDate")
    List<Task> getInDeadline();

    @Query("SELECT * FROM tasks ORDER BY title")
    List<Task> getInNameAsc();

    @Query("SELECT * FROM tasks " +
            "WHERE category=:category AND startDate BETWEEN :startDate AND :endDate")
    List<Task> getByFilter(int category, long startDate, long endDate);

    @Query("SELECT * FROM tasks " +
            "WHERE startDate BETWEEN :startDate AND :endDate")
    List<Task> getByFilter(long startDate, long endDate);
}

File: app/src/main/java/cedule/app/data/dao/TaskDAO.kt
package cedule.app.data.dao;

import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.*
import cedule.app.data.entities.Task
import kotlinx.coroutines.flow.Flow

@Dao
interface TaskDAO {
    @Query("SELECT * FROM tasks")
    fun getAll(): Flow<List<Task>>

    @Query("SELECT * FROM tasks WHERE category=:category")
    fun getByCategory(category: Int): Flow<List<Task>>

    @Update
    suspend fun update(task: Task)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun add(task: Task)

    @Query("SELECT * FROM tasks WHERE id=:id LIMIT 1")
    fun getById(id: Int): Flow<Task>

    @Query("DELETE FROM tasks WHERE id IN (:id)")
    suspend fun discard(id: List<Int>)

    @Query("UPDATE tasks SET isDone=:isDone WHERE id=:id")
    suspend fun updateStatus(id: Int, isDone: Int)

    @Query("SELECT * FROM tasks WHERE startDate == :startDate")
    fun getByDate(startDate: Long): Flow<List<Task>>

    @Transaction
    suspend fun insertOrUpdate(task: Task) {
        if (task.id != null) update(task) else add(task)
    }

    @Query("SELECT * FROM tasks ORDER BY createTime DESC LIMIT 1")
    suspend fun getLatestTask(): Task?

    @Query("SELECT COUNT(*) FROM tasks WHERE startDate=:date AND isDone=0")
    fun countTasks(date: Long): Flow<Int>
}